In [1]:
import pandas as pd
import numpy as np

In [2]:
data=pd.read_csv('C:\\Users\\RAM\\Downloads\\03. Pandas\\Datasets/employee.csv')
data.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

###### By two ways we can check for null values in our whole dataset

In [5]:
data.isnull().any().any()

False

In [6]:
data.isnull().sum().sum()

0

In [8]:
#we have no null values in the employee dataset, Let's take another dataset and work on it

data = pd.read_csv('C:\\Users\\RAM\\Downloads\\03. Pandas\\Datasets/melbourne.csv')
data.head(3)

Suburb          Address  Rooms Type      Price Method SellerG  \
0  Abbotsford    68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford     85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford  25 Bloomburg St      2    h  1035000.0      S  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  

[3 rows x 21 columns]

In [9]:
data.isnull().sum().sum()

66918

In [10]:
data.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

In [11]:
# first lets check how many rows are having all the null values
data.isnull().all(axis = 1).sum()

0

In [12]:
# lets check the percentage of missing values in all the columns

percentage_of_missing_values = round(100*(data.isnull().sum()/len(data.index)), 2)
print(percentage_of_missing_values)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64


In [13]:
# removing the three columns having 33% to 56% of missing values in the dataset
data = data.drop('BuildingArea', axis=1)
data = data.drop('YearBuilt', axis=1)
data = data.drop('CouncilArea', axis=1)

In [14]:
# count the number of rows having > 5 missing values
# use len(df.index)
len(data[data.isnull().sum(axis=1) > 5].index)

4278

In [15]:
# retaining the rows having <= 5 NaNs
data = data[data.isnull().sum(axis=1) <= 5]

In [16]:
# look at the summary again
round(100*(data.isnull().sum()/len(data.index)), 2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.71
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2          1.05
Bathroom          1.07
Car               1.81
Landsize          9.65
Lattitude         0.13
Longtitude        0.13
Regionname        0.00
Propertycount     0.00
dtype: float64

It can be seen that Price Column still has large number of missing values i.e., 21% of the data is missing. If we impute these values it will introduce heavy bias into the data and the results will be misleading so It is better to discard the rows where Price is having missing values.

In [17]:
# removing all the rows where there is a missing value in the Price Column
data = data[~np.isnan(data['Price'])]

In [18]:
# look at the summary again
round(100*(data.isnull().sum()/len(data.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         1.05
Bathroom         1.07
Car              1.76
Landsize         9.83
Lattitude        0.15
Longtitude       0.15
Regionname       0.00
Propertycount    0.00
dtype: float64

* Now, we have Landsize, which is having high percentage of missing values, lets check the range of values present in the Landsize attribute of the data, so that we can impute the missing values present in the Landsize Attribute.

In [20]:
data['Landsize'].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

* From the above description of the range of values present in the landsize columns supported by mean, standard deviation, minimum, maximum, 25% percentile, 50% percentile, 75% percentile etc.

* We can see that the Minimum value is 0 and Maximum value is 433014, there is huge difference so both mean and median function would not work

* Also, there is a huge difference in the 25th, 50th, and 75th percentiles indicating that if we impute the values for Landsize column we will most probably introduce bias into the data.

* It is most appropriate to remove the rows where we find null values for Landsize.

In [21]:
# removing  all the rows where there is a null value in the landsize column
data = data[~np.isnan(data['Landsize'])]

# lets check the summary again
round(100*(data.isnull().sum()/len(data.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.46
Landsize         0.00
Lattitude        0.16
Longtitude       0.16
Regionname       0.00
Propertycount    0.00
dtype: float64

In [22]:
# describing the values for lattitude and longitude
data[['Lattitude','Longtitude']].describe()

Lattitude    Longtitude
count  13581.000000  13581.000000
mean     -37.809204    144.995221
std        0.079257      0.103913
min      -38.182550    144.431810
25%      -37.856820    144.929600
50%      -37.802360    145.000100
75%      -37.756400    145.058320
max      -37.408530    145.526350

In [23]:
# as there is minute difference between the values of min, mean, max, 25th, 50th, and 75th percentile. 
# so we can use mean function to impute the values in lattitude and longtitude.

data['Lattitude'].fillna(data['Lattitude'].mean(), inplace = True)
data['Longtitude'].fillna(data['Longtitude'].mean(), inplace = True)

In [24]:
# lets check the values present in the Bathroom Column

data['Bathroom'].value_counts()

1.0    7517
2.0    4987
3.0     921
4.0     106
0.0      34
5.0      28
6.0       5
8.0       2
7.0       2
Name: Bathroom, dtype: int64

In [25]:
# lets also check the values present in the Car Column

data['Car'].value_counts()

2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: Car, dtype: int64

In [26]:
# we have to impute the values in Categorical Columns using the Mode function

data['Bathroom'].fillna(data['Bathroom'].mode()[0], inplace = True)
data['Car'].fillna(data['Car'].mode()[0], inplace = True)

In [27]:
data.isnull().sum().sum()

0